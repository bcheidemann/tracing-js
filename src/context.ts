/**
 * @module
 * This module provides functions and type definitions to work with the AsyncLocalStorage tracing context.
 * It's primarily intended for internal use, but the API is public, and may be useful when creating custom
 * subscribers.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import type { ISubscriber } from "./subscriber.ts";

export type Context = {
  subscriber: ISubscriber<unknown>;
  clone(): Context;
};

export const context: AsyncLocalStorage<Context> = new AsyncLocalStorage<
  Context
>();

export function createContext(subscriber: ISubscriber<unknown>): Context {
  return {
    subscriber,
    clone() {
      return createContext(this.subscriber.clone());
    },
  };
}

export function getContextOrThrow(): Context {
  const ctx = context.getStore();

  if (!ctx) {
    throw new Error("No context found");
  }

  return ctx;
}

export function getContext(): Context | undefined {
  return context.getStore();
}
