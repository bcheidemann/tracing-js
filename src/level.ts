/**
 * @module
 * This module provides the Level type and enum.
 */

export interface Levels {
  TRACE: 0;
  DEBUG: 1;
  INFO: 2;
  WARN: 3;
  ERROR: 4;
  CRITICAL: 5;
  DISABLED: 6;
}

export type Level = Levels[keyof Levels];

export const Level = {
  DISABLED: 0,
  TRACE: 1,
  DEBUG: 2,
  INFO: 3,
  WARN: 4,
  ERROR: 5,
  CRITICAL: 6,
} as const;
