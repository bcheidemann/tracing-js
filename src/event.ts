/**
 * @module
 * This module provides functions for creating events.
 */

import { getSubscriberContext } from "./context.ts";
import { Level } from "./level.ts";

/**
 * The Event type represents the information which is passed to subscribers when an event is created.
 */
export type Event = {
  isEvent: true;
  level: Level;
  message: string;
  fields?: Record<string, unknown>;
};

/**
 * Creates a new event and passes it to the current subscriber.
 *
 * @param level The level of the event
 * @param message The message of the event
 * @param fields The fields of the event (optional additional information)
 */
export function event(
  level: Level,
  message: string,
  fields?: Record<string, unknown>,
): void {
  const ctx = getSubscriberContext();

  if (!ctx) {
    return;
  }

  if (
    ctx.subscriber.enabledForLevel &&
    !ctx.subscriber.enabledForLevel(level)
  ) {
    return;
  }

  const event: Event = {
    isEvent: true,
    level,
    message,
    fields,
  };

  if (
    ctx.subscriber.enabled &&
    !ctx.subscriber.enabled(event)
  ) {
    return;
  }

  ctx.subscriber.event({
    isEvent: true,
    level,
    message,
    fields,
  });
}

/**
 * Creates a new trace level event and passes it to the current subscriber.
 *
 * @param message The message of the event
 * @param fields The fields of the event (optional additional information)
 */
export function trace(message: string, fields?: Record<string, unknown>) {
  event(Level.TRACE, message, fields);
}

/**
 * Creates a new debug level event and passes it to the current subscriber.
 *
 * @param message The message of the event
 * @param fields The fields of the event (optional additional information)
 */
export function debug(message: string, fields?: Record<string, unknown>) {
  event(Level.DEBUG, message, fields);
}

/**
 * Creates a new info level event and passes it to the current subscriber.
 *
 * @param message The message of the event
 * @param fields The fields of the event (optional additional information)
 */
export function info(message: string, fields?: Record<string, unknown>) {
  event(Level.INFO, message, fields);
}

/**
 * Creates a new warn level event and passes it to the current subscriber.
 *
 * @param message The message of the event
 * @param fields The fields of the event (optional additional information)
 */
export function warn(message: string, fields?: Record<string, unknown>) {
  event(Level.WARN, message, fields);
}

/**
 * Creates a new error level event and passes it to the current subscriber.
 *
 * @param message The message of the event
 * @param fields The fields of the event (optional additional information)
 */
export function error(message: string, fields?: Record<string, unknown>) {
  event(Level.ERROR, message, fields);
}

/**
 * Creates a new critical level event and passes it to the current subscriber.
 *
 * @param message The message of the event
 * @param fields The fields of the event (optional additional information)
 */
export function critical(message: string, fields?: Record<string, unknown>) {
  event(Level.CRITICAL, message, fields);
}

/**
 * The Log object provides functions for creating events. It is a convenience object which provides the same functions
 * as the module itself.
 */
export const Log = {
  trace,
  debug,
  info,
  warn,
  error,
  critical,
};
