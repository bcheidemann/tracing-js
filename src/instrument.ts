import { parseParamNamesFromFunction } from "@bcheidemann/parse-params";
import { context, getContext } from "./context.ts";
import { event } from "./event.ts";
import { Level } from "./level.ts";
import { span } from "./span.ts";
import { AssertionError } from "node:assert";

const AttributeKind = {
  Message: 0 as const,
  Target: 1 as const,
  Level: 2 as const,
  Skip: 3 as const,
  SkipAll: 4 as const,
  SkipThis: 5 as const,
  Field: 6 as const,
  LogEnter: 7 as const,
  LogExit: 8 as const,
  LogError: 9 as const,
  Log: 10 as const,
};

type AttributeKind = typeof AttributeKind;

type MessageAttribute = {
  kind: AttributeKind["Message"];
  message: string;
};

type TargetAttribute = {
  kind: AttributeKind["Target"];
} & (
  | {
      class: string;
      method: string;
      private?: boolean;
    }
  | {
      function: string;
    }
);

type LevelAttribute = {
  kind: AttributeKind["Level"];
  level: Level;
};

type SkipByMask<TArgs extends any[]> = {
  [index in keyof TArgs]: boolean;
};
type SkipAttribute<TArgs extends any[]> = {
  kind: AttributeKind["Skip"];
  skip: SkipByMask<TArgs> | string[] | number[];
};

type SkipAllAttribute = {
  kind: AttributeKind["SkipAll"];
};

type SkipThisAttribute = {
  kind: AttributeKind["SkipThis"];
};

type FieldAttribute<TArgs extends any[]> = {
  kind: AttributeKind["Field"];
  name: string;
  value: unknown | ((args: TArgs[]) => unknown);
};

type LogEnterAttribute = {
  kind: AttributeKind["LogEnter"];
};

type LogExitAttribute = {
  kind: AttributeKind["LogExit"];
};

type LogErrorAttribute = {
  kind: AttributeKind["LogError"];
};

type LogAttribute = {
  kind: AttributeKind["Log"];
};

type Attributes<TArgs extends any[]> =
  | MessageAttribute
  | TargetAttribute
  | LevelAttribute
  | SkipAttribute<TArgs>
  | SkipAllAttribute
  | SkipThisAttribute
  | FieldAttribute<TArgs>
  | LogEnterAttribute
  | LogExitAttribute
  | LogErrorAttribute
  | LogAttribute;

type InstrumentDecorator<TMethod extends (this: any, ...args: any[]) => any> = (
  target: TMethod,
  context: ClassMethodDecoratorContext<ThisParameterType<TMethod>, TMethod>,
) => TMethod;

export function message(message: string): MessageAttribute {
  return { kind: AttributeKind.Message, message };
}

export function target(
  className: string,
  method: string,
  isPrivate?: boolean,
): TargetAttribute;
export function target(functionName: string): TargetAttribute;
export function target(
  classOrFunctionName: string,
  method?: string,
  isPrivate?: boolean,
): TargetAttribute {
  if (method !== undefined) {
    return {
      kind: AttributeKind.Target,
      class: classOrFunctionName,
      method,
      private: isPrivate ?? false,
    };
  }
  return { kind: AttributeKind.Target, function: classOrFunctionName };
}

export function level(level: Level): LevelAttribute {
  return { kind: AttributeKind.Level, level };
}

export function skip<TArgs extends any[]>(
  ...mask: SkipByMask<TArgs>
): SkipAttribute<TArgs>;
export function skip(...paramNames: string[]): SkipAttribute<any>;
export function skip(...paramIndices: number[]): SkipAttribute<any>;
export function skip<TArgs extends any[]>(
  ...skip: SkipAttribute<TArgs>["skip"]
): SkipAttribute<TArgs> {
  return { kind: AttributeKind.Skip, skip };
}

export const skipAll: SkipAllAttribute = { kind: AttributeKind.SkipAll };

export const skipThis: SkipThisAttribute = { kind: AttributeKind.SkipThis };

export function field<TArgs extends any[]>(
  name: string,
  mapValue: (args: NoInfer<TArgs>) => unknown,
): FieldAttribute<TArgs>;
export function field<TArgs extends any[], TValue>(
  name: string,
  value: TValue extends (...args: any[]) => any ? never : TValue,
): FieldAttribute<TArgs>;
export function field<TArgs extends any[]>(
  name: string,
  value: unknown | ((...args: TArgs[]) => unknown),
): FieldAttribute<TArgs> {
  return { kind: AttributeKind.Field, name, value };
}

export const logEnter: LogEnterAttribute = { kind: AttributeKind.LogEnter };

export const logExit: LogExitAttribute = { kind: AttributeKind.LogExit };

export const logError: LogErrorAttribute = { kind: AttributeKind.LogError };

export const log: LogAttribute = { kind: AttributeKind.Log };

export function instrument<TMethod extends (this: any, ...args: any[]) => any>(
  ...attributes: Attributes<Parameters<TMethod>>[]
): InstrumentDecorator<TMethod> {
  return function instrumentDecorator(
    target: TMethod,
    ctx: ClassMethodDecoratorContext<ThisParameterType<TMethod>, TMethod>,
  ) {
    return instrumentCallbackImpl(target, attributes, {
      kind: "method",
      isPrivate: ctx.private,
      methodName: typeof ctx.name === "symbol" ? ctx.name.toString() : ctx.name,
    });
  };
}

export function instrumentCallback<
  TCallback extends (this: any, ...args: any[]) => any,
>(fn: TCallback): TCallback;
export function instrumentCallback<
  TCallback extends (this: any, ...args: any[]) => any,
>(attributes: Attributes<Parameters<TCallback>>[], fn: TCallback): TCallback;
export function instrumentCallback<
  TCallback extends (this: any, ...args: any[]) => any,
>(
  attributesOrFn: TCallback | Attributes<Parameters<TCallback>>[],
  fn?: TCallback,
): TCallback {
  if (fn) {
    return instrumentCallbackImpl(
      fn,
      attributesOrFn as Attributes<Parameters<TCallback>>[],
      { kind: "function" },
    );
  }

  return instrumentCallbackImpl(attributesOrFn as TCallback, [], {
    kind: "function",
  });
}

type Defaults = {
  [AttributeKind.Message]: MessageAttribute;
  [AttributeKind.Target]: TargetAttribute;
  [AttributeKind.Level]: LevelAttribute;
};

function collectAttributes<TArgs extends any[]>(
  attributes: Attributes<TArgs>[],
  defaults: Defaults,
): {
  [AttributeKind.Message]: MessageAttribute;
  [AttributeKind.Target]: TargetAttribute;
  [AttributeKind.Level]: LevelAttribute;
  [AttributeKind.Skip]: SkipAttribute<TArgs>[];
  [AttributeKind.SkipAll]?: SkipAllAttribute;
  [AttributeKind.SkipThis]?: SkipThisAttribute;
  [AttributeKind.Field]: FieldAttribute<TArgs>[];
  [AttributeKind.LogEnter]?: LogEnterAttribute;
  [AttributeKind.LogExit]?: LogExitAttribute;
  [AttributeKind.LogError]?: LogErrorAttribute;
  [AttributeKind.Log]?: LogAttribute;
} {
  const attributesByKind: ReturnType<typeof collectAttributes<TArgs>> = {
    [AttributeKind.Message]: defaults[AttributeKind.Message],
    [AttributeKind.Target]: defaults[AttributeKind.Target],
    [AttributeKind.Level]: defaults[AttributeKind.Level],
    [AttributeKind.Skip]: [],
    [AttributeKind.SkipAll]: undefined,
    [AttributeKind.SkipThis]: undefined,
    [AttributeKind.Field]: [],
    [AttributeKind.LogEnter]: undefined,
    [AttributeKind.LogExit]: undefined,
    [AttributeKind.LogError]: undefined,
    [AttributeKind.Log]: undefined,
  };

  for (const attribute of attributes) {
    switch (attribute.kind) {
      case AttributeKind.Message:
        attributesByKind[AttributeKind.Message] = attribute;
        break;
      case AttributeKind.Target:
        attributesByKind[AttributeKind.Target] = attribute;
        break;
      case AttributeKind.Level:
        attributesByKind[AttributeKind.Level] = attribute;
        break;
      case AttributeKind.Skip:
        attributesByKind[AttributeKind.Skip].push(attribute);
        break;
      case AttributeKind.SkipAll:
        attributesByKind[AttributeKind.SkipAll] = attribute;
        break;
      case AttributeKind.SkipThis:
        attributesByKind[AttributeKind.SkipThis] = attribute;
        break;
      case AttributeKind.Field:
        attributesByKind[AttributeKind.Field].push(attribute);
        break;
      case AttributeKind.LogEnter:
        attributesByKind[AttributeKind.LogEnter] = attribute;
        break;
      case AttributeKind.LogExit:
        attributesByKind[AttributeKind.LogExit] = attribute;
        break;
      case AttributeKind.LogError:
        attributesByKind[AttributeKind.LogError] = attribute;
        break;
      case AttributeKind.Log:
        attributesByKind[AttributeKind.Log] = attribute;
        break;
      default:
        const _: never = attribute;
        throw new AssertionError({
          message: "Invalid attribute kind",
          actual: (attribute as any).kind,
          expected: Object.keys(AttributeKind).join(" | "),
        });
    }
  }

  return attributesByKind;
}

type Context =
  | {
      kind: "function";
    }
  | {
      kind: "method";
      isPrivate: boolean;
      methodName: string;
    };

const ParsedFunctionParams = new WeakMap<Function, string[]>();

function instrumentCallbackImpl<
  TCallback extends (this: any, ...args: any[]) => any,
>(
  fn: TCallback,
  attributes: Attributes<Parameters<TCallback>>[],
  instrumentCtx: Context,
): TCallback {
  return function instrumentedCallback(
    this: ThisParameterType<TCallback>,
    ...args: Parameters<TCallback>
  ) {
    const fnName = fn.name || "[unknown function]";
    const defaults: Defaults = {
      [AttributeKind.Message]:
        instrumentCtx.kind === "function"
          ? message(fn.name)
          : message(
              `${this?.constructor?.name ?? "[unknown class]"}.${fnName}`,
            ),
      [AttributeKind.Target]:
        instrumentCtx.kind === "function"
          ? target(fnName)
          : target(
              this?.constructor?.name ?? "[unknown class]",
              instrumentCtx.methodName,
              instrumentCtx.isPrivate,
            ),
      [AttributeKind.Level]: level(Level.INFO),
    };
    const attributesByKind = collectAttributes(attributes, defaults);
    const self = this;
    const ctx = getContext().clone();
    return context.run(ctx, function () {
      const level = attributesByKind[AttributeKind.Level].level;
      const message = attributesByKind[AttributeKind.Message].message;
      const target = (() => {
        const targetAttribute = attributesByKind[AttributeKind.Target];
        if ("class" in targetAttribute) {
          return {
            class: targetAttribute.class,
            method: targetAttribute.method,
            private: targetAttribute.private,
          };
        }
        return {
          function: targetAttribute.function,
        };
      })();
      const logArgs = (() => {
        const logArgs: Record<number, unknown> & { this?: unknown } = {
          ...args,
        };
        if (attributesByKind[AttributeKind.SkipAll] !== undefined) {
          return [];
        }
        if (attributesByKind[AttributeKind.SkipThis] === undefined) {
          logArgs["this"] = self;
        }
        for (const skipAttribute of attributesByKind[AttributeKind.Skip]) {
          skipAttribute.skip.forEach((skip, index) => {
            switch (typeof skip) {
              case "string":
                const paramNames =
                  ParsedFunctionParams.get(fn) ??
                  parseParamNamesFromFunction(fn);
                ParsedFunctionParams.set(fn, paramNames);
                const skipIndex = paramNames.indexOf(skip);
                delete logArgs[skipIndex];
                break;
              case "number":
                delete logArgs[skip];
                break;
              case "boolean":
                if (skip) {
                  delete logArgs[index];
                }
                break;
              default:
                throw new AssertionError({
                  message: "Invalid type for skip attribute value",
                  actual: typeof skip,
                  expected: "string | number | boolean",
                });
            }
          });
        }
        return logArgs;
      })();
      const log = Boolean(attributesByKind[AttributeKind.Log]);
      const logEnter = log || Boolean(attributesByKind[AttributeKind.LogEnter]);
      const logExit = log || Boolean(attributesByKind[AttributeKind.LogExit]);
      const logError = log || Boolean(attributesByKind[AttributeKind.LogError]);
      const fields = attributesByKind[AttributeKind.Field].reduce(
        (acc, field) => {
          acc[field.name] =
            typeof field.value === "function" ? field.value(args) : field.value;
          return acc;
        },
        {} as Record<string, unknown>,
      );
      const guard = span(level, message, {
        target,
        args: logArgs,
        ...fields,
      }).enter();
      try {
        if (logEnter) {
          event(level, `Entering ${message}`);
        }
        const returnValue = fn.apply(self, args);

        // Handle async error / success
        if (returnValue instanceof Promise) {
          return returnValue
            .catch((error) => {
              if (logError) {
                event(level, `Error in ${message}`, { error });
              }
              guard.exit();
              throw error;
            })
            .then((returnValue) => {
              if (logExit) {
                event(level, `Exiting ${message}`, { returnValue });
              }
              guard.exit();
              return returnValue;
            });
        }

        // Handle sync success
        if (logExit) {
          event(level, `Exiting ${message}`, { returnValue });
        }
        guard.exit();
        return returnValue;
      } catch (err) {
        // Handle sync errors
        if (logError) {
          event(level, `Error in ${message}`, { error: err });
        }
        guard.exit();
        throw err;
      }
    });
  } as TCallback;
}
