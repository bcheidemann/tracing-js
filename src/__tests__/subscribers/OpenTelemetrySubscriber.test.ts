import { beforeEach, describe, it } from "@std/testing/bdd";
import { expect } from "expect";
import { event } from "../../event.ts";
import { span } from "../../span.ts";
import { Level } from "../../level.ts";
import { getSubscriberContext } from "../../context.ts";
import { instrumentCallback, message } from "../../instrument.ts";
import { OpenTelemetrySubscriber } from "../../subscribers/OpenTelemetrySubscriber.ts";
import { InMemorySpanProcessor } from "../otel.ts";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import { trace } from "@opentelemetry/api";

describe("OpenTelemetrySubscriber", () => {
  const spanProcessor = new InMemorySpanProcessor();
  const provider = new NodeTracerProvider({
    spanProcessors: [spanProcessor],
  });
  provider.register();

  const tracer = trace.getTracer("test");

  beforeEach(() => {
    spanProcessor.reset();
  });

  it("should record a span without events", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });

    // Act
    const guard = span(Level.INFO, "test span").enter();
    guard.exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(1);
    expect(spans[0].name).toEqual("test span");
    expect(spans[0].events).toHaveLength(0);
  });

  it("should record a span with attributes", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });

    // Act
    const guard = span(Level.INFO, "test span", {
      stringAttr: "example string",
      numberAttr: 42,
      booleanAttr: true,
      nullAttr: null,
      bigintAttr: BigInt(42),
      functionAttr: function exampleFunction() {},
      classAttr: class ExampleClass {},
      dateAttr: new Date(0),
      errorAttr: new Error("example error"),
      instanceAttr: new class ExampleClass {
        public message = "Hello World!";
      }(),
      objectAttr: {
        title: "Greeting",
        message: "Hello World!",
      },
      nestedObjectAttr: {
        type: "email",
        content: {
          html: "<h1>Hello World!</h1>",
          text: "Hello World!",
        },
      },
      mixedArrayAttr: ["example string", 42, true, null],
      emptyArrayAttr: [],
      stringArrayAttr: [null, "1", "2", "3"],
      numberArrayAttr: [null, 1, 2, 3],
      booleanArrayAttr: [null, true, false, true],
    }).enter();
    guard.exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(1);
    expect(spans[0].name).toEqual("test span");
    expect(spans[0].attributes).toBeDefined();
    expect(spans[0].attributes).toMatchObject({
      stringAttr: "example string",
      numberAttr: 42,
      booleanAttr: true,
      nullAttr: "null",
      bigintAttr: "42",
      functionAttr: "function exampleFunction() {}",
      classAttr: expect.stringContaining("class ExampleClass"),
      dateAttr: "1970-01-01T00:00:00.000Z",
      "errorAttr.name": "Error",
      "errorAttr.message": "example error",
      "errorAttr.stack": expect.stringContaining("Error: example error\n"),
      instanceAttr: "ExampleClass",
      "instanceAttr.message": "Hello World!",
      "objectAttr.title": "Greeting",
      "objectAttr.message": "Hello World!",
      "nestedObjectAttr.type": "email",
      "nestedObjectAttr.content.html": "<h1>Hello World!</h1>",
      "nestedObjectAttr.content.text": "Hello World!",
      mixedArrayAttr: ["example string", "42", "true", null],
      emptyArrayAttr: [],
      stringArrayAttr: [null, "1", "2", "3"],
      numberArrayAttr: [null, 1, 2, 3],
      booleanArrayAttr: [null, true, false, true],
    });
    expect(spans[0].events).toHaveLength(0);
  });

  it("should record a nested span without events", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });

    // Act
    const outerGuard = span(Level.INFO, "outer span").enter();
    const innerGuard = span(Level.INFO, "inner span").enter();
    innerGuard.exit();
    outerGuard.exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(2);
    expect(spans[0].name).toEqual("inner span");
    expect(spans[0].events).toHaveLength(0);
    expect(spans[1].name).toEqual("outer span");
    expect(spans[1].events).toHaveLength(0);
    expect(spans[0].parentSpanContext?.spanId).toEqual(
      spans[1].spanContext().spanId,
    );
  });

  it("should record events on a span", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });

    // Act
    const guard = span(Level.INFO, "test span").enter();
    event(Level.INFO, "test event");
    guard.exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(1);
    expect(spans[0].name).toEqual("test span");
    expect(spans[0].events).toHaveLength(1);
    expect(spans[0].events[0].name).toEqual("test event");
  });

  it("should record events with attributes on a span", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });

    // Act
    const guard = span(Level.INFO, "test span").enter();
    event(Level.INFO, "test event", {
      stringAttr: "example string",
      numberAttr: 42,
      booleanAttr: true,
      nullAttr: null,
      bigintAttr: BigInt(42),
      functionAttr: function exampleFunction() {},
      classAttr: class ExampleClass {},
      dateAttr: new Date(0),
      errorAttr: new Error("example error"),
      instanceAttr: new class ExampleClass {
        public message = "Hello World!";
      }(),
      objectAttr: {
        title: "Greeting",
        message: "Hello World!",
      },
      nestedObjectAttr: {
        type: "email",
        content: {
          html: "<h1>Hello World!</h1>",
          text: "Hello World!",
        },
      },
      mixedArrayAttr: ["example string", 42, true, null],
      emptyArrayAttr: [],
      stringArrayAttr: [null, "1", "2", "3"],
      numberArrayAttr: [null, 1, 2, 3],
      booleanArrayAttr: [null, true, false, true],
    });
    guard.exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(1);
    expect(spans[0].name).toEqual("test span");
    expect(spans[0].events).toHaveLength(1);
    expect(spans[0].events[0].name).toEqual("test event");
    expect(spans[0].events[0].attributes).toMatchObject({
      stringAttr: "example string",
      numberAttr: 42,
      booleanAttr: true,
      nullAttr: "null",
      bigintAttr: "42",
      functionAttr: "function exampleFunction() {}",
      classAttr: expect.stringContaining("class ExampleClass"),
      dateAttr: "1970-01-01T00:00:00.000Z",
      "errorAttr.name": "Error",
      "errorAttr.message": "example error",
      "errorAttr.stack": expect.stringContaining("Error: example error\n"),
      instanceAttr: "ExampleClass",
      "instanceAttr.message": "Hello World!",
      "objectAttr.title": "Greeting",
      "objectAttr.message": "Hello World!",
      "nestedObjectAttr.type": "email",
      "nestedObjectAttr.content.html": "<h1>Hello World!</h1>",
      "nestedObjectAttr.content.text": "Hello World!",
      mixedArrayAttr: ["example string", "42", "true", null],
      emptyArrayAttr: [],
      stringArrayAttr: [null, "1", "2", "3"],
      numberArrayAttr: [null, 1, 2, 3],
      booleanArrayAttr: [null, true, false, true],
    });
  });

  it("should record events on a nested span", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });

    // Act
    const guard = span(Level.INFO, "outer span").enter();
    const innerGuard = span(Level.INFO, "inner span").enter();
    event(Level.INFO, "test event");
    innerGuard.exit();
    guard.exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(2);
    expect(spans[0].name).toEqual("inner span");
    expect(spans[0].events).toHaveLength(1);
    expect(spans[0].events[0].name).toEqual("test event");
    expect(spans[1].name).toEqual("outer span");
    expect(spans[1].events).toHaveLength(0);
    expect(spans[0].parentSpanContext?.spanId).toEqual(
      spans[1].spanContext().spanId,
    );
  });

  it("should record fields on the span before it is entered", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });
    const newSpan = span(Level.INFO, "test span");

    // Act
    newSpan.record("test1", "testA").record("test2", "testB").enter().exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(1);
    expect(spans[0].name).toEqual("test span");
    expect(spans[0].attributes).toMatchObject(
      expect.objectContaining({
        test1: "testA",
        test2: "testB",
      }),
    );
  });

  it("should record fields on the span after it is entered", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });
    const newSpan = span(Level.INFO, "test span");

    // Act
    newSpan.enter().record("test1", "testA").record("test2", "testB").exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(1);
    expect(spans[0].name).toEqual("test span");
    expect(spans[0].attributes).toMatchObject(
      expect.objectContaining({
        test1: "testA",
        test2: "testB",
      }),
    );
  });

  it("should record fields on the span before and after it is entered", () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });
    const newSpan = span(Level.INFO, "test span");

    // Act
    newSpan.record("beforeEnter", "A").enter().record("afterEnter", "B").exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(1);
    expect(spans[0].name).toEqual("test span");
    expect(spans[0].attributes).toMatchObject(
      expect.objectContaining({
        beforeEnter: "A",
        afterEnter: "B",
      }),
    );
  });

  it("should handle concurrent contexts correctly", async () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });
    async function first() {
      const inner = span(Level.INFO, "first inner").enter();
      await new Promise((resolve) => setTimeout(resolve, 10));
      event(Level.INFO, "first");
      await new Promise((resolve) => setTimeout(resolve, 20));
      inner.exit();
    }
    async function second() {
      const inner = span(Level.INFO, "second inner").enter();
      await new Promise((resolve) => setTimeout(resolve, 20));
      event(Level.INFO, "second");
      await new Promise((resolve) => setTimeout(resolve, 20));
      inner.exit();
    }

    // Act
    const outerGuard = span(Level.INFO, "outer").enter();
    await Promise.all([
      instrumentCallback([message("first function")], first)(),
      instrumentCallback([message("second function")], second)(),
    ]);
    outerGuard.exit();

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    expect(spans).toHaveLength(5);

    expect(spans[0].name).toEqual("first inner");
    expect(spans[0].events).toHaveLength(1);
    expect(spans[0].events[0].name).toEqual("first");
    expect(spans[1].name).toEqual("first function");
    expect(spans[1].events).toHaveLength(0);

    expect(spans[2].name).toEqual("second inner");
    expect(spans[2].events).toHaveLength(1);
    expect(spans[2].events[0].name).toEqual("second");
    expect(spans[3].name).toEqual("second function");
    expect(spans[3].events).toHaveLength(0);

    expect(spans[4].name).toEqual("outer");
    expect(spans[4].events).toHaveLength(0);

    expect(spans[0].parentSpanContext?.spanId).toEqual(
      spans[1].spanContext().spanId,
    );
    expect(spans[1].parentSpanContext?.spanId).toEqual(
      spans[4].spanContext().spanId,
    );

    expect(spans[2].parentSpanContext?.spanId).toEqual(
      spans[3].spanContext().spanId,
    );
    expect(spans[3].parentSpanContext?.spanId).toEqual(
      spans[4].spanContext().spanId,
    );
  });

  it.skip("should handle delayed execution on inner spans", async () => {
    // Arrange
    OpenTelemetrySubscriber.setGlobalDefault({ tracer });

    const { promise: testCompletion, resolve: completeTest } = Promise
      .withResolvers<true>();

    const delayedFn = instrumentCallback(async function delayedFn() {
      const inner = span(Level.INFO, "inner").enter();
      await new Promise((resolve) => setTimeout(resolve, 10));
      event(Level.INFO, "delayed event");
      await new Promise((resolve) => setTimeout(resolve, 20));
      inner.exit();

      // Complete the test in the next event loop cycle
      setTimeout(() => completeTest(true));
    });

    const main = instrumentCallback(function main() {
      // TODO: This is a bit of a hack and only preserves the OTEL span
      //       heirarchy... We need to provide some API with which to bind a
      //       callback to the current tracing span, even if the current span
      //       exits before the callback executes.
      getSubscriberContext()!.subscriber.runInContext(
        () => {
          setTimeout(() => delayedFn(), 10);
        },
        null,
        [],
      );
    });

    // Act
    main();
    await testCompletion;

    // Assert
    const spans = spanProcessor.getFinishedSpans();
    const processedSpans = spans.map((span) => ({
      name: span.name,
      id: span.spanContext().spanId,
      parentId: span.parentSpanContext?.spanId,
    }));
    console.error(JSON.stringify(processedSpans, null, 2));
    expect(spans).toHaveLength(3);

    expect(spans[0].name).toEqual("main");
    expect(spans[0].events).toHaveLength(0);
    expect(spans[1].name).toEqual("inner");
    expect(spans[1].events).toHaveLength(1);
    expect(spans[1].events[0].name).toEqual("delayed event");
    expect(spans[2].name).toEqual("delayedFn");
    expect(spans[2].events).toHaveLength(0);

    expect(spans[1].parentSpanContext?.spanId).toEqual(
      spans[2].spanContext().spanId,
    );
    expect(spans[2].parentSpanContext?.spanId).toEqual(
      spans[0].spanContext().spanId,
    );
  });
});
