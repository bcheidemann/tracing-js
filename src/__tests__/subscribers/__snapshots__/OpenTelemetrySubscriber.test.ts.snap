export const snapshot = {};

snapshot[`OpenTelemetrySubscriber > should log warnings for incorrectly exited spans 1`] = `
"[WARNING] (OpenTelemetrySubscriber) A span was exited before its child.

          [EXITED]
          name = outer
          level = 3

          [ORPHANS]
          0: name = inner
             level = 3

          This probably means that a function was called asynchronously
          but was not properly instrumented.

          To fix this error, ensure that asynchronously executed
          functions are properly instrumented.

          For example:

          setTimeout(instrumentCallback(() => { ... }));
                     +++++++++++++++++++             +

          const myFunc = instrumentCallback(async function myFync() { ... });
          ++++++++++++++++++++++++++++++++++                               ++

          class Example {
            @instrument()
            +++++++++++++
            myMethod() { ... }
          }
"
`;

snapshot[`OpenTelemetrySubscriber > should log warnings for incorrectly exited spans 2`] = `
"[WARNING] (OpenTelemetrySubscriber) A span was exited before its child.

          [EXITED]
          name = outer
          level = 3

          [ORPHANS]
          0: name = inner
             level = 3

          This probably means that a function was called asynchronously
          but was not properly instrumented.

          To fix this error, ensure that asynchronously executed
          functions are properly instrumented.

          For example:

          setTimeout(instrumentCallback(() => { ... }));
                     +++++++++++++++++++             +

          const myFunc = instrumentCallback(async function myFync() { ... });
          ++++++++++++++++++++++++++++++++++                               ++

          class Example {
            @instrument()
            +++++++++++++
            myMethod() { ... }
          }
"
`;

snapshot[`OpenTelemetrySubscriber > should log warnings for exiting a span which does not belong to the current subscriber context 1`] = `
"[WARNING] (OpenTelemetrySubscriber) A span was exited which does not
          belong to the current subscribers context. This probably means
          that one of the following happened:

          [1] The spans _id property was mutated. This is an invalid
              operation.

          [2] The OpenTelemetrySubscriber.exit method was called
              manually with an invalid span ID.
"
`;
